src/main/java/com/squareup/javapoet/CodeWriter.java

```diff
@@ -272,106 +272,80 @@ private void emitLiteral(Object o) throws IOException {
+    for (ClassName c = className; c != null; c = c.enclosingClassName()) {
+      ClassName resolved = resolve(c.simpleName());
+      nameResolved = resolved != null;
+
+      if (Objects.equals(resolved, c)) {
+        int suffixOffset = c.simpleNames().size() - 1;
+        return join(".", className.simpleNames().subList(
+            suffixOffset, className.simpleNames().size()));
       }
+    }
 
-      // Fall back to the fully-qualified name. Mark the type as importable for a future pass.
-      if (!javadoc) importableTypes.add(className);
-      return className.canonicalName;
+    // If the class is in the same package, we're done.
+    if (Objects.equals(packageName, className.packageName())) {
+      referencedNames.add(className.topLevelClassName().simpleName());
+      return join(".", className.simpleNames());
     }
 
-    // Look for the longest non-conflicting common prefix, which we can omit.
-    List<String> classNames = className.simpleNames();
-    int prefixLength = nonConflictingCommonPrefixLength(classNames);
-    if (prefixLength == classNames.size()) {
-      return className.simpleName(); // Special case: a class referring to itself!
+    // We'll have to use the fully-qualified name. Mark the type as importable for a future pass.
+    if (!javadoc && !nameResolved) {
+      importableType(className);
     }
 
-    referencedNames.add(classNames.get(0));
-    return Util.join(".", classNames.subList(prefixLength, classNames.size()));
+    return className.canonicalName;
   }
 
-  /**
-   * Returns true if {@code className} conflicts with a visible class name in the current scope and
-   * cannot be referred to by its short name.
-   */
-  private boolean conflictsWithLocalName(ClassName className) {
-    for (TypeSpec typeSpec : typeSpecStack) {
-      if (Objects.equals(typeSpec.name, className.simpleName())) return true;
-      for (TypeSpec visibleChild : typeSpec.typeSpecs) {
-        if (Objects.equals(visibleChild.name, className.simpleName())) return true;
-      }
+  private void importableType(ClassName className) {
+    ClassName topLevelClassName = className.topLevelClassName();
+    String simpleName = topLevelClassName.simpleName();
+    ClassName replaced = importableTypes.put(simpleName, topLevelClassName);
+    if (replaced != null) {
+      importableTypes.put(simpleName, replaced); // On collision, prefer the first inserted.
     }
-    return false;
   }
 
   /**
-   * Returns the common prefix of {@code classNames} and the current nesting scope. For example,
-   * suppose the current scope is {@code AbstractMap.SimpleEntry}. This will return 0 for {@code
-   * List}, 1 for {@code AbstractMap}, 1 for {@code AbstractMap.SimpleImmutableEntry}, and 2 for
-   * {@code AbstractMap.SimpleEntry} itself.
+   * Returns the class referenced by {@code simpleNames}, using the current nesting context and
+   * imports.
+   *
+   * TODO(jwilson): also honor superclass members when resolving names.
```

Comment:
`//`-style comment? Otherwise this shows up in Javadoc and people's IDEs. Unless you're going to fix before release.